cpp/src/structs/search/contains.cu

```diff
@@ -35,52 +35,33 @@ bool contains(structs_column_view const& haystack,
+  // comparator.
+  auto const comparator = cudf::experimental::row::equality::two_table_comparator(
+    haystack_tv, table_view{{needle_as_col}}, stream);
+  auto const has_nulls = has_nested_nulls(haystack_tv) || has_nested_nulls(needle_tv);
+  auto const d_comp    = comparator.device_comparator(nullate::DYNAMIC{has_nulls});
 
-  // Prepare to flatten the structs column and scalar.
-  auto const has_null_elements = has_nested_nulls(table_view{std::vector<column_view>{
-                                   haystack.child_begin(), haystack.child_end()}}) ||
-                                 has_nested_nulls(scalar_table);
-  auto const flatten_nullability = has_null_elements
-                                     ? structs::detail::column_nullability::FORCE
-                                     : structs::detail::column_nullability::MATCH_INCOMING;
-
-  // Flatten the input structs column, only materialize the bitmask if there is null in the input.
-  auto const haystack_flattened =
-    structs::detail::flatten_nested_columns(table_view{{haystack}}, {}, {}, flatten_nullability);
-  auto const needle_flattened =
-    structs::detail::flatten_nested_columns(scalar_table, {}, {}, flatten_nullability);
-
-  // The struct scalar only contains the struct member columns.
-  // Thus, if there is any null in the input, we must exclude the first column in the flattened
-  // table of the input column from searching because that column is the materialized bitmask of
-  // the input structs column.
-  auto const haystack_flattened_content  = haystack_flattened.flattened_columns();
-  auto const haystack_flattened_children = table_view{std::vector<column_view>{
-    haystack_flattened_content.begin() + static_cast<size_type>(has_null_elements),
-    haystack_flattened_content.end()}};
-
-  auto const d_haystack_children_ptr =
-    table_device_view::create(haystack_flattened_children, stream);
-  auto const d_needle_ptr = table_device_view::create(needle_flattened, stream);
-
-  auto const start_iter = thrust::make_counting_iterator<size_type>(0);
+  auto const start_iter = cudf::experimental::row::lhs_iterator(0);
   auto const end_iter   = start_iter + haystack.size();
-  auto const comp       = row_equality_comparator(nullate::DYNAMIC{has_null_elements},
-                                            *d_haystack_children_ptr,
-                                            *d_needle_ptr,
-                                            null_equality::EQUAL);
+  using cudf::experimental::row::rhs_index_type;
+
   auto const found_iter = thrust::find_if(
-    rmm::exec_policy(stream), start_iter, end_iter, [comp] __device__(auto const idx) {
-      return comp(idx, 0);  // compare haystack[idx] == val[0].
+    rmm::exec_policy(stream), start_iter, end_iter, [d_comp] __device__(auto const idx) {
+      // Compare needle_as_col[0] == haystack[idx].
+      return d_comp(static_cast<rhs_index_type>(0), idx);
```

Comment:
Since the result is symmetric, let's call this in the `left, right` order. That eliminates the need to dispatch through the `right, left` operator to the `left, right` operator.
```suggestion
      return d_comp(idx, static_cast<rhs_index_type>(0));
```

Reply:
The dispatch is inline and is optimized out. I intentionally switched the side of the operand to trigger the calling to `operator()(rhs_, lhs_)` so we can test it.

Reply:
This is fine for testing but we should call in `left, right` order in the final state of this PR.

Reply:
Okay done.